contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,116 @@
+* Regional data costs in multi-AZ setup - for instance, in AWS, with a multi-AZ setup, half of the traffic will switch AZ, incurring regional data Transfer costs, whereas if something was local, it wouldn't hit the network.
+* Users want a priority-based topology aware routing  service. Say, try to visit a "host-local" service backend firstly. If no available backend, then try a "zone-local" service backend...
+
+## Background
+
+It has been a very often requested feature because people need to create a service which connects to local backend Pods, the so called "only-local" service. In this way, they can reduce network latency, improve security, save money and so on. However, because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to enable "local-only service".
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+## Implementation details
+
+### API changes
+
+Create a new `TopologyKeyList` filed in `Service.Spec`.
+
+```go
+type Service struct {
+  Spec:
+    // topologyKey is the key for the node label that the system uses to denote such a topology domain. There are some built-in topology keys, e.g. kubernetes.io/hostname,
+    // failure-domain.beta.kubernetes.io/zone,
+    // failure-domain.beta.kubernetes.io/region,
+    // TopologyKeyList is a topologyKey list, which is used to guarantee ability. For example,
+    // `TopologyKeyList: [ "kubernetes.io/hostname", "failure-domain.beta.kubernetes.io/zone" ]`
+    // If failed to match hostname, then it will try to find the endpoints in the same zone
+    // The priority of key in TopologyKeyList is sorted by index and selecting process will stop once a key matches.
+  	TopologyKeyList []string
+}
+```
+
+Although `NodeName` was already added to `EndpointAddress`, we want `Endpoints` to carry more node's topology informations so that allowing more topology levels other than hostname. 
+
+So, create a new `NodeLabels` field in `Endpoints.Subsets.Addresses` for identifying what topology domain the endpoints pod exists, e.g. what host, rack, zone, region etc. In other words, copy the labels of node hosting the endpoint to `EndpointAddress.NodeLabels`.
```

Comment:
how do we keep the node labels (for rack, zone, region) in sync with the label endpoint NodeLabels, will the service controller be responsible for this ?

Reply:
endpoint controller is expected to do the sync.